File: kotlin/src/test/kotlin/OcrTest.kt
import Ocr.parse
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test

class OcrTest {
    @Test
    fun `parses one record with 888888888`() {
        assertEquals(
            listOf("888888888    "),
            parse(
                " _   _   _   _   _   _   _   _   _  ",
                "|_| |_| |_| |_| |_| |_| |_| |_| |_| ",
                "|_| |_| |_| |_| |_| |_| |_| |_| |_| ",
                "                                    "
            )
        )
    }

    @Test
    fun `parses two records with 888888888`() {
        assertEquals(
            listOf(
                "888888888    ",
                "888888888    "
            ),
            parse(
                " _   _   _   _   _   _   _   _   _  ",
                "|_| |_| |_| |_| |_| |_| |_| |_| |_| ",
                "|_| |_| |_| |_| |_| |_| |_| |_| |_| ",
                "                                    ",
                " _   _   _   _   _   _   _   _   _  ",
                "|_| |_| |_| |_| |_| |_| |_| |_| |_| ",
                "|_| |_| |_| |_| |_| |_| |_| |_| |_| ",
                "                                    "
            )
        )
    }

    @Test
    fun `parses one record with 123456790`() {
        assertEquals(
            listOf("123456790    "),
            parse(
                "     _   _       _   _   _   _   _  ",
                "  |  _|  _| |_| |_  |_    | |_| | | ",
                "  | |_   _|   |  _| |_|   |  _| |_| ",
                "                                    "
            )
        )
    }

    @Test
    fun `parses two records`() {
        assertEquals(
            listOf(
                "123456790    ",
                "947874222    "
            ),
            parse(
                "     _   _       _   _   _   _   _  ",
                "  |  _|  _| |_| |_  |_    | |_| | | ",
                "  | |_   _|   |  _| |_|   |  _| |_| ",
                "                                    ",
                " _       _   _   _       _   _   _  ",
                "|_| |_|   | |_|   | |_|  _|  _|  _| ",
                " _|   |   | |_|   |   | |_  |_  |_  ",
                "                                    "
            )
        )
    }

    @Test
    fun `parses illegal digit`() {
        assertEquals(
            listOf("12??56790 ILL"),
            parse(
                "     _   _       _   _   _   _   _  ",
                "  |  _| |_|  _| |_  |_    | |_| | | ",
                "  | |_   _    |  _| |_|   |  _| |_| ",
                "                                    "
            )
        )
    }

//    @Test
//    fun `checksum fail returns ERR`() {
//        assertEquals(
//            listOf("664371495 ERR"),
//            parse(
//                " _   _       _   _           _   _  ",
//                "|_  |_  |_|  _|   |   | |_| |_| |_  ",
//                "|_| |_|   |  _|   |   |   |  _|  _| ",
//                "                                    "
//            )
//        )
//    }
}

File: php/test/OcrTest.php
<?php

namespace Tests;

use PHPUnit\Framework\TestCase;
use Ocr\Ocr;

class OcrTest extends TestCase {

    /**
     * @test
     */
    public function parses_one_record_with_888888888() {
        $expected = ["888888888    "];
        $actual = Ocr::parse(" _   _   _   _   _   _   _   _   _  ",
                             "|_| |_| |_| |_| |_| |_| |_| |_| |_| ",
                             "|_| |_| |_| |_| |_| |_| |_| |_| |_| ",
                             "                                    ");
        $this->assertEquals($expected, $actual);
    }

    /**
     * @test
     */
    public function parses_two_records_with_888888888() {
        $expected = ["888888888    ", 
                     "888888888    "];
        $actual = Ocr::parse(" _   _   _   _   _   _   _   _   _  ",
                             "|_| |_| |_| |_| |_| |_| |_| |_| |_| ",
                             "|_| |_| |_| |_| |_| |_| |_| |_| |_| ",
                             "                                    ",
                             " _   _   _   _   _   _   _   _   _  ",
                             "|_| |_| |_| |_| |_| |_| |_| |_| |_| ",
                             "|_| |_| |_| |_| |_| |_| |_| |_| |_| ",
                             "                                    ");
        $this->assertEquals($expected, $actual);
    }

    /**
     * @test
     */
    public function parses_one_record_with_123456790() {
        $expected = ["123456790    "];
        $actual = Ocr::parse("     _   _       _   _   _   _   _  ",
                             "  |  _|  _| |_| |_  |_    | |_| | | ",
                             "  | |_   _|   |  _| |_|   |  _| |_| ",
                             "                                    ");
        $this->assertEquals($expected, $actual);
    }

    /**
     * @test
     */
    public function parses_two_records() {
        $expected = ["123456790    ", 
                     "947874222    "];
        $actual = Ocr::parse("     _   _       _   _   _   _   _  ",
                             "  |  _|  _| |_| |_  |_    | |_| | | ",
                             "  | |_   _|   |  _| |_|   |  _| |_| ",
                             "                                    ",
                             " _       _   _   _       _   _   _  ",
                             "|_| |_|   | |_|   | |_|  _|  _|  _| ",
                             " _|   |   | |_|   |   | |_  |_  |_  ",
                             "                                    ");
        $this->assertEquals($expected, $actual);
    }

    /**
     * @test
     */
    public function parses_illegal_digit() {
        $expected = ["12??56790 ILL"];
        $actual = Ocr::parse("     _   _       _   _   _   _   _  ",
                             "  |  _| |_|  _| |_  |_    | |_| | | ",
                             "  | |_   _    |  _| |_|   |  _| |_| ",
                             "                                    ");
        $this->assertEquals($expected, $actual);
    }

    // /**
    //  * @test
    //  */
    // public function checksum_fail_returns_ERR() {
    //     $expected = ["664371495 ERR"];
    //     $actual = Ocr::parse(" _   _       _   _           _   _  ",
    //                          "|_  |_  |_|  _|   |   | |_| |_| |_  ",
    //                          "|_| |_|   |  _|   |   |   |  _|  _| ",
    //                          "                                    ");
    //     $this->assertEquals($expected, $actual);
    // }

}
